node/beacon-chain-handler/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+extern crate beacon;
+extern crate chain;
+extern crate node_runtime;
+extern crate primitives;
+extern crate storage;
+
+use beacon::types::BeaconBlock;
+use chain::BlockChain;
+use node_runtime::{ApplyState, Runtime};
+use primitives::traits::{Block, Header};
+use primitives::traits::Signer;
+use primitives::types::ConsensusBlockBody;
+use primitives::types::SignedTransaction;
+use std::sync::Arc;
+use storage::StateDb;
+
+pub trait ConsensusHandler<B: Block, P>: Send + Sync {
+    fn produce_block(&self, body: ConsensusBlockBody<P>) -> B;
+}
+
+pub struct BeaconBlockProducer {
```

Comment:
seems like we split Client into 3 objects (BeaconBlockProducer, StateDbViewer and whatever future object that will import blocks?) which have exactly the same content as `Client` had but are across different cargos?

Reply:
they don't need to be split up between crates but i wanted to just start extracting things from client. i agree we should do some crate consolidation in the future. it seems like we are maybe abusing crates in general and should lean towards modules in some places.

as to why i am splitting them up: they don't have the same content. `BlockImporter` may be implemented with the producer.. not sure yet.

```
pub struct StateDbViewer {
    beacon_chain: Arc<BlockChain<BeaconBlock>>,
    state_db: Arc<StateDb>,
}
```

```
pub struct BeaconBlockProducer {
    beacon_chain: Arc<BlockChain<BeaconBlock>>,
    runtime: Runtime,
    signer: Arc<Signer>,
    state_db: Arc<StateDb>,
}
```

part of the motivation behind separating the client is to eliminate a god object being relied upon everywhere. another part is to eventually get to a point where we have interfaces that are copyable without worrying about write locks across threads. i think we are stuck on blockchain's and state db's, but have some ideas for how to solve elegantly without locking in the future (like providing eventual consistency on reads from the `StateDbViewer`). @nearmax may be able to comment more.